node/network/availability-distribution/src/requester/fetch_task.rs

```diff
@@ -0,0 +1,397 @@
+	{
+		let FetchTaskConfig {
+			prepared_running,
+			live_in,
+		} = config;
+
+		if let Some(running) = prepared_running {
+			let (handle, kill) = oneshot::channel();
+
+			ctx.spawn("chunk-fetcher", running.run(kill).boxed())
+				.await
+				.map_err(|e| Error::SpawnTask(e))?;
+
+			Ok(FetchTask {
+				live_in,
+				state: FetchedState::Started(handle),
+			})
+		} else {
+			Ok(FetchTask {
+				live_in,
+				state: FetchedState::Canceled,
+			})
+		}
+	}
+
+	/// Add the given leaf to the relay parents which are making this task relevant.
+	///
+	/// This is for book keeping, so we know we are already fetching a given chunk.
+	pub fn add_leaf(&mut self, leaf: Hash) {
+		self.live_in.insert(leaf);
+	}
+
+	/// Remove leaves and cancel the task, if it was the last one and the task has still been
+	/// fetching.
+	pub fn remove_leaves(&mut self, leaves: &HashSet<Hash>) {
+		self.live_in.difference(leaves);
+		if self.live_in.is_empty() && !self.is_finished() {
+			self.state = FetchedState::Canceled
+		}
+	}
+
+	/// Whether or not there are still relay parents around with this candidate pending
+	/// availability.
+	pub fn is_live(&self) -> bool {
+		!self.live_in.is_empty()
+	}
+
+	/// Whether or not this task can be considered finished.
+	///
+	/// That is, it is either canceled, succeeded or failed.
+	pub fn is_finished(&self) -> bool {
+		match &self.state {
+			FetchedState::Canceled => true,
+			FetchedState::Started(sender) => sender.is_canceled(),
+		}
+	}
+}
+
+/// Things that can go wrong in task execution.
+#[derive(Debug)]
+enum TaskError {
+	/// The peer failed to deliver a correct chunk for some reason (has been reported as
+	/// appropriate).
+	PeerError,
+	/// This very node is seemingly shutting down (sending of message failed).
+	ShuttingDown,
+}
+
+impl RunningTask {
+	async fn run(self, kill: oneshot::Receiver<()>) {
+		// Wait for completion/or cancel.
+		let run_it = self.run_inner();
+		futures::pin_mut!(run_it);
+		let _ = select(run_it, kill).await;
+	}
+
+	/// Fetch and store chunk.
+	///
+	/// Try validators in backing group in order.
+	async fn run_inner(mut self) {
+		let mut bad_validators = Vec::new();
+		// Try validators in reverse order:
+		while let Some(validator) = self.group.pop() {
+			// Send request:
+			let resp = match self.do_request(&validator).await {
+				Ok(resp) => resp,
+				Err(TaskError::ShuttingDown) => {
+					tracing::info!(
+						target: LOG_TARGET,
+						"Node seems to be shutting down, canceling fetch task"
+					);
+					return
+				}
+				Err(TaskError::PeerError) => {
+					bad_validators.push(validator);
+					continue
+				}
+			};
+			let chunk = match resp {
+				AvailabilityFetchingResponse::Chunk(resp) => {
+					resp.reconstruct_erasure_chunk(&self.request)
+				}
+				AvailabilityFetchingResponse::NoSuchChunk => {
+					tracing::debug!(target: LOG_TARGET, validator = ?validator, "Validator did not have our chunk");
+					bad_validators.push(validator);
```

Comment:
I don't think we should consider validators who simply don't have the data as on the same level as those who provide incorrect data. It is possible that some validators in the group just don't have it.

But the punishment for being a 'bad_validator` isn't that high now, just means they're not requested from as much, so I won't block on this.

Reply:
Yeah, this is just from a networking perspective - we did not get what we wanted, so we just wasted our time - try someone else next time.

I would think providing an invalid chunk would be a candidate for slashing?

Reply:
The ideal thing to do would be to request from the actual _backers_ but we don't expose that via `OccupiedCore` right now. Also there would need to be some unification of backer bitfield (bitwise OR) because different subsets of backers might appear in different chains and we'd want to request from anyone who signed a statement. Future work, but please file an issue.

Reply:
Another thing to do would be to send notifications about having the data to all validators. And then validators can build up a list of other validators who claim to have the data rather than requesting from arbitrary others.

Reply:
https://github.com/paritytech/polkadot/issues/2509